compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -620,6 +620,55 @@ impl<'a> Parser<'a> {
             );
         }
 
+        // Possible missing `struct` or `enum` keyword
+        if let TokenKind::Ident(symbol, _) = &self.prev_token.kind
+            && let TokenKind::OpenDelim(Delimiter::Brace) = self.token.kind
+            && let [TokenType::Token(token::Not), TokenType::Token(token::PathSep)] = expected[..]
+        {
+            let mut lookahead = self.clone();
+            for position in 1..1000 {
+                lookahead.bump();
+                if lookahead.token == token::CloseDelim(Delimiter::Brace)
+                    || lookahead.token == token::Colon
+                {
+                    let (article, kw) = if lookahead.token == token::Colon || position == 1 {
+                        ("a", "struct")
+                    } else {
+                        ("an", "enum")
+                    };
+                    err.span_suggestion(
+                        self.prev_token.span,
+                        format!("if this is {article} {kw} definition, use the `{kw}` keyword"),
+                        format!("{kw} {symbol}"),
+                        Applicability::MaybeIncorrect,
+                    );
+                    break;
+                }
+            }
+        }
+
+        // Possible missing `fn` keyword
+        if let TokenKind::Ident(symbol, _) = &self.prev_token.kind
+            && let TokenKind::OpenDelim(Delimiter::Parenthesis) = self.token.kind
+        {
+            let mut lookahead = self.clone();
+            for _ in 1..100 {
```

Comment:
I think we should just leave recovery alone here. Eating up to 100 tokens before a semicolon seems really arbitrary imo, and doesn't really have to do with the suggestion. I assume it's to suppress *other* suggestions -- that can be done separately, perhaps by implementing a better recovery in `parse_fn_front_matter` rather than here.

Reply:
I don't think this is recovery, it's just checking that we hit an open brace before we hit a semicolon, to prevent the suggestion firing on `main();` in item position, and the 100 is an arbitrary bound so as not to do unbounded lookahead. Is there a different approach you'd like me to try?

Reply:
Probably moving this logic completely out of `expected_one_of_not_found` and into something like `parse_item` (or something near there in the call stack).

Reply:
Yeah, I second the suggestion to move this from `expected_one_of_not_found` to somewhere else near functions handling items.